streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -630,57 +631,188 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
-            queue.take.flatten.foldCause(
-              ZChannel.refailCause,
-              {
-                case Left(outDone)  => ZChannel.succeedNow(outDone)
-                case Right(outElem) => ZChannel.write(outElem) *> consumer
-              }
+    mapOutZIOPar[Env1, OutErr1, OutElem2](n, n)(f)
+
+  final def mapOutZIOPar[Env1 <: Env, OutErr1 >: OutErr, OutElem2](n: Int, bufferSize: Int)(
+    f: OutElem => ZIO[Env1, OutErr1, OutElem2]
+  )(implicit trace: Trace): ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone] = {
+    val z: ZIO[Any, Nothing, ZChannel[Env1, InErr, InElem, InDone, OutErr1, OutElem2, OutDone]] = for {
+      input         <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
+      queueReader    = ZChannel.fromInput(input)
+      queue         <- Queue.bounded[Fiber[Option[OutErr1], OutElem2]](bufferSize)
+      permits       <- zio.Semaphore.make(n)
+      failureSignal <- Promise.make[Option[OutErr1], Nothing]
+      outDoneSignal <- Promise.make[Nothing, OutDone]
+    } yield {
+      def forkF(a: OutElem): ZIO[Env1, Nothing, Fiber.Runtime[Option[OutErr1], OutElem2]] = ZIO.uninterruptibleMask {
+        restore =>
+          for {
+            localScope <- zio.Scope.make
+            _          <- restore(permits.withPermitScoped.provideEnvironment(ZEnvironment(localScope)))
+            fib <- restore {
+                     f(a) //.debug("forkF.f(a)")
+                       .foldCauseZIO(
+                         c => failureSignal.failCause(c.map(Some(_))) *> ZIO.refailCause(c.map(Some(_))),
+                         u => ZIO.succeed(u)
+                       )
+                       //.debug("forkF.folded")
+                       .raceWith[Env1, Option[OutErr1], Option[OutErr1], Nothing, OutElem2](failureSignal.await)(
+                         { case (leftEx, rightFib) =>
+                           rightFib.interrupt *> leftEx
+                         },
+                         { case (rightEx, leftFib) =>
+                           leftFib.interrupt *> rightEx
+                         }
+                       )
+                     //.debug("forkF.race")
+                   }
+                     .onExit(localScope.close(_))
+                     .fork
+          } yield fib
+      }
+
+      lazy val enqueueCh: ZChannel[Env1, OutErr, OutElem, OutDone, Nothing, Nothing, Any] = ZChannel
+        .readWithCause(
+          in => ZChannel.fromZIO(forkF(in).flatMap(queue.offer(_))) *> enqueueCh,
+          err =>
+            ZChannel.fromZIO(
+              failureSignal.failCause(err.map(Some(_))) *> queue.offer(Fiber.failCause(err.map(Some(_))))
+            ),
+          done => ZChannel.fromZIO(outDoneSignal.succeed(done) *> queue.offer(Fiber.fail(None)))
+        )
+
+      val enqueuer: ZIO[Env1 with Scope, Nothing, Fiber.Runtime[Nothing, Any]] = queueReader
+        .pipeTo(self)
+        .pipeTo(enqueueCh)
+        .runScoped
+        .forkScoped
```

Comment:
yes, the channel is running in its own fiber which is kept alive with some management code that somehow binds the fiber's lifetime to the scope's lifetime. this is important since the optimized code relies on its spawned fibers to be bound to the channel's lifetime (specifically the enqueuer channel+fiber)

Reply:
see [this](https://github.com/zio/zio/blob/series/2.x/streams/shared/src/main/scala/zio/stream/ZChannel.scala#L1074), the channel is executed within its own fiber and this fiber is kept alive until the scope is closed.
all fibers spawned by effects executed by the streams are descendants of this fiber (unless explicitly forked as daemons or explicitly `transplant`ed)